### PR TITLE
Refactoring fetch_bookmark to avoid extra queries

### LINKED ISSUE #806: `fetch_bookmark` does more queries than we'd like when `get_direct_children` is true
One option is to modify `fetch_tree` so that it's usable in this case. Follow-up from #743

┆Issue is synchronized with this [Jira Story](https://jira.allizom.org/browse/SYNC-128)

### PR DESCRIPTION
**Pull request by lyuyuan92**
fixes #806

### CODE REVIEW

**Comment by lyuyuan92** [comment]
Hi @thomcc, it seems `shasum -a 256 -c openssl-1.1.1a.tar.gz` failed in the Rust benchmarks test. Could you help me rerun this test? It doesn't look like the failure is related to my changes. Thanks!

**Comment by thomcc** [comment]
Retriggered, I agree it seems unrelated.

Just FYI, most of the team is traveling for a conference so review turnaround might be slightly slower this week.

File: components/places/src/storage/bookmarks.rs [hunk 1]
```diff
@@ -1178,34 +1184,79 @@ pub fn fetch_tree(
...
+                }
+                .into(),
             }
-            .into()
         }
         None => return Ok(None),
     };
 
+    // Skip the rest and return if root is not a folder
+    match root {
```
[the remark this answers is not in the document]
Comment (lyuyuan92) [reply components/places/src/storage/bookmarks.rs]: Done.

File: components/places/src/storage/bookmarks.rs [hunk 2]
```diff
@@ -1178,34 +1184,79 @@ pub fn fetch_tree(
...
     // For all remaining rows, build a pseudo-tree that maps parent GUIDs to
     // ordered children. We need this intermediate step because SQLite returns
     // results in level order, so we'll see a node's siblings and cousins (same
     // level, but different parents) before any of their descendants.
     let mut pseudo_tree: HashMap<SyncGuid, Vec<BookmarkTreeNode>> = HashMap::new();
+
+    // Keep track of folders' guid of each level. Initialize the vector with two
+    // sets, which are for the root and its immediate children folders.
+    let mut node_vec: Vec<HashSet<SyncGuid>> = vec![HashSet::new(); 2];
+    let mut curr_depth: usize = 0;
```
[the remark this answers is not in the document]
Comment (lyuyuan92) [reply components/places/src/storage/bookmarks.rs]: Ah, I am a bit silly here...totally forgot about the `level` field. Done.

File: components/places/src/storage/bookmarks/public_node.rs [hunk 3]
```diff
@@ -93,105 +93,38 @@ pub fn fetch_bookmark(
...
-    if parent.bookmark_type != BookmarkType::Folder {
-        return Ok(ChildInfo::NoChildren);
-    }
-    // If we already know that we have no children, don't
-    // bother querying for them.
-    if parent.child_count == 0 {
-        return Ok(if get_direct_children {
-            ChildInfo::Nodes(vec![])
+fn set_child_guids_and_drop(parent: &mut PublicNode) {
+    if let Some(children) = parent.child_nodes.take() {
```
[the remark this answers is not in the document]
Comment (lyuyuan92) [reply components/places/src/storage/bookmarks/public_node.rs]: Done. The suggested code is much cleaner, thanks!

**Pull request merged**